File: week4&5/Ariling/boj-2623-음악프로그램/main.py
import sys
from collections import deque
input = sys.stdin.readline

[N, M] = list(map(int, input().split()))
indegrees = [0] * (N+1)
graph = [[] for _ in range(N+1)]
for _ in range(M):
    infoArr = list(map(int, input().split()))
    if len(infoArr) > 2:
        for i in range(1, len(infoArr)-1):
            graph[infoArr[i]].append(infoArr[i+1])
            indegrees[infoArr[i+1]] += 1
q = deque([])
result = []
for i in range(1, len(indegrees)):
    if indegrees[i] == 0:
        q.append(i)
        result.append(i)
while q:
    current = q.popleft()
    for i in range(len(graph[current])):
        next = graph[current][i]
        indegrees[next] -= 1
        if indegrees[next] == 0:
            q.append(next)
            result.append(next)
if len(result) == N:
    for num in result:
        print(num)
else:
    print(0)

File: week4&5/Ariling/boj-2623-음악프로그램/main.js
const fs = require("fs");
const [firstline, ...input] = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");
// 이거 순서가 있는건데 분명히... 이것도 위상정렬인 것 같은데...?
const info_arr = firstline.split(" ").map(Number);
const pd_info_arr = input.map((line) => line.split(" ").map(Number));
function solution(info_arr, pd_info_arr) {
  const graph = Array(info_arr[0] + 1)
    .fill()
    .map(() => []);
  const indegree = Array(info_arr[0] + 1).fill(0);
  const result_arr = [];
  for (let i = 0; i < pd_info_arr.length; i++) {
    const [cnt, ...info] = pd_info_arr[i];
    if (cnt > 1) {
      for (let j = 0; j < cnt - 1; j++) {
        graph[info[j]].push(info[j + 1]);
        indegree[info[j + 1]]++;
      }
    }
  }

  let q = [];
  for (let i = 1; i <= info_arr[0]; i++) {
    if (indegree[i] === 0) {
      q.push(i);
      result_arr.push(i);
    }
  }
  while (q.length > 0) {
    const current = q.shift();
    for (let i = 0; i < graph[current].length; i++) {
      const next = graph[current][i];
      indegree[next]--;

      if (indegree[next] === 0) {
        q.push(next); // 진입차수 0이면 추가해주기
        result_arr.push(next);
      }
    }
  }
  // 출력하기
  if (result_arr.length === info_arr[0]) {
    for (const el of result_arr) {
      console.log(el);
    }
  } else {
    console.log(0);
  }
}

solution(info_arr, pd_info_arr);
